src/vector/backends/awkward.py

```diff
@@ -1978,16 +1978,85 @@ def impl(record: typing.Any) -> typing.Any:
 
 
-ak.behavior["__numba_typer__", "Vector2D"] = _numba_typer_Vector2D
-ak.behavior["__numba_typer__", "Vector3D"] = _numba_typer_Vector3D
-ak.behavior["__numba_typer__", "Vector4D"] = _numba_typer_Vector4D
-ak.behavior["__numba_typer__", "Momentum2D"] = _numba_typer_Momentum2D
-ak.behavior["__numba_typer__", "Momentum3D"] = _numba_typer_Momentum3D
-ak.behavior["__numba_typer__", "Momentum4D"] = _numba_typer_Momentum4D
-
-ak.behavior["__numba_lower__", "Vector2D"] = _numba_lower
-ak.behavior["__numba_lower__", "Vector3D"] = _numba_lower
-ak.behavior["__numba_lower__", "Vector4D"] = _numba_lower
-ak.behavior["__numba_lower__", "Momentum2D"] = _numba_lower
-ak.behavior["__numba_lower__", "Momentum3D"] = _numba_lower
-ak.behavior["__numba_lower__", "Momentum4D"] = _numba_lower
+behavior["__numba_typer__", "Vector2D"] = _numba_typer_Vector2D
+behavior["__numba_typer__", "Vector3D"] = _numba_typer_Vector3D
+behavior["__numba_typer__", "Vector4D"] = _numba_typer_Vector4D
+behavior["__numba_typer__", "Momentum2D"] = _numba_typer_Momentum2D
+behavior["__numba_typer__", "Momentum3D"] = _numba_typer_Momentum3D
+behavior["__numba_typer__", "Momentum4D"] = _numba_typer_Momentum4D
+
+behavior["__numba_lower__", "Vector2D"] = _numba_lower
+behavior["__numba_lower__", "Vector3D"] = _numba_lower
+behavior["__numba_lower__", "Vector4D"] = _numba_lower
+behavior["__numba_lower__", "Momentum2D"] = _numba_lower
+behavior["__numba_lower__", "Momentum3D"] = _numba_lower
+behavior["__numba_lower__", "Momentum4D"] = _numba_lower
+
+
+def _reduce_sum(
+    array: VectorArray2D
+    | VectorArray3D
+    | VectorArray4D
+    | MomentumArray2D
+    | MomentumArray3D
+    | MomentumArray4D,
+    mask_identity: bool,
+) -> VectorProtocol:
+    fields = {}
+    if isinstance(array, Lorentz):
+        fields["t"] = numpy.sum(array.t, axis=1)
+    if isinstance(array, Spatial):
+        fields["z"] = numpy.sum(array.z, axis=1)
+
+    assert isinstance(array, Planar)
+    fields["x"] = numpy.sum(array.x, axis=1)
+    fields["y"] = numpy.sum(array.y, axis=1)
+
+    layout = ak.to_layout(array)
+
+    return ak.zip(
+        fields,
+        behavior=array.behavior,
+        with_name=layout.purelist_parameter("__record__"),
+    )
+
+
+def _reduce_count(
+    array: VectorArray2D
+    | VectorArray3D
+    | VectorArray4D
+    | MomentumArray2D
+    | MomentumArray3D
+    | MomentumArray4D,
+    mask_identity: bool,
+) -> VectorProtocol:
+    first_field = array[array.fields[0]]
+    return ak.count(first_field, axis=1)
+
+
+def _reduce_count_nonzero(
+    array: VectorArray2D
+    | VectorArray3D
+    | VectorArray4D
+    | MomentumArray2D
+    | MomentumArray3D
+    | MomentumArray4D,
+    mask_identity: bool,
+) -> ScalarCollection:
+    mag_2 = array.rho2
+    if isinstance(array, Spatial):
+        mag_2 = mag_2 + array.z**2
+    if isinstance(array, Lorentz):
+        mag_2 = mag_2 + array.t2
+
+    return ak.count_nonzero(mag_2, axis=1)
```

Comment:
Instead of checking that the sum is equal to zero, this could check to see if all components are individually equal to zero (three `== 0` and two `&` on the `rho2`, `z`, and `t2` components). That _feels_ cleaner, though I don't see any way that floating point errors in `+` could cause accidental cancellations in rare cases: all three of these expressions are guaranteed to be positive. It even gives the same result if any are `NaN`. (`NaN` counts as nonzero and `NaN` is infectious: it will spread through the addition.)

So this is okay.

Reply:
I initially thought about implementing it that way. Off the top of my head, using the sum is slightly fewer vectorised operations (`array == 0` and 3 `__add__` vs a single `pow(x, 2)` and 3 `__add__`, for Lorentz). It was late when I wrote this, though, because floating-point addition is slower than integer addition and that's enough to make the aesthetics matter too.

On balance, I prefer your suggestion.